### PR TITLE
Basic persistence for blueprints

### PR DESCRIPTION
**Pull request by jleibs**
### What
When moving over to the new blueprint store, we temporarily lost the ability to persist blueprints.

This is one of the major regressions currently preventing us from releasing main (See: #2529)

This PR adds new methods to the `store_hub`:
 - `persist_app_blueprints`
 - `try_to_load_persisted_blueprint`
 
 The blueprints are stored using the standard rrd file format. They go into a sub-directory of our app-state storage:
- Linux: /home/UserName/.local/share/rerun/blueprints/
- macOS: /Users/UserName/Library/Application Support/rerun/blueprints/
- Windows: C:\Users\UserName\AppData\Roaming\rerun\blueprints\

We only store a single blueprint in this folder per app-id. 

Any time we change the app_id via the `store_hub` (such as when loading an rrd) we will search the directory for a persisted blueprint and then load it. We track the `insert_id` on the active blueprint to determine if we need to save back out to disk again. I've tested to confirm that in most cases when we restore from blueprint we don't do anything weird like re-modify the blueprint and save it again.

### Additional Notes:
The way that DataBlueprint/AutoValues work required a workaround that I don't particularly love, but does seems to be workable until we refactor to get rid of more of the serde business.
   - After implementing blueprint-loading, I noticed that every time we start the app fresh, we see a cycle where the AutoValue for `pinhole_image_plane_distance` goes from Auto(100.06751) -> Auto(1) ->  Auto(100.03121). Even ignoring the transient jump due to view bounds being a frame delayed to populate, the Auto-value itself is unstable from run to run.  This meant that every time we started the app, we would do a new insertion which would cause the blueprint to immediately get saved again.
   - The work-around is to replace the usage of `!=` (PartialEq) with a conceptually equivalent method `has_edits` for the DataBlueprintTree.  We now consider all values of Auto to be considered a match. This seems to work out fine since we (should?) always recompute Auto values at the beginning of each frame.
   - Longer term, I think the right solution is to move the Auto values out of the blueprint and into the AppState so they persist frame-to-frame and don't even show up in the blueprint comparison logic. However, DataBlueprintTree is a complicated piece of work, and plumbing through that state is proving to be challenging.
   
### Future work
This currently only implements file-backed storage, so doesn't work with web-view.
 - https://github.com/rerun-io/rerun/issues/2579

### Testing:
* Run colmap fiat:
```
python examples/python/structure_from_motion/main.py --dataset colmap_fiat
```
* Adjust views
* Exit application
* Check the md5sum of the blueprint:
```
$ md5sum /home/jleibs/.local/share/rerun/blueprints/structure_from_motion.blueprint 
03fab330d9c23ac3da1163c854aa8518  /home/jleibs/.local/share/rerun/blueprints/structure_from_motion.blueprint
```
* Run colmap fiat a *second* time.
* Confirm the layout persists:
image
* Close rerun and verify the blueprint hasn't been changed:
```
$ md5sum /home/jleibs/.local/share/rerun/blueprints/structure_from_motion.blueprint 
03fab330d9c23ac3da1163c854aa8518  /home/jleibs/.local/share/rerun/blueprints/structure_from_motion.blueprint
```

### Checklist
* [x] I have read and agree to [Contributor Guide](https://github.com/rerun-io/rerun/blob/main/CONTRIBUTING.md) and the [Code of Conduct](https://github.com/rerun-io/rerun/blob/main/CODE_OF_CONDUCT.md)
* [x] I've included a screenshot or gif (if applicable)
* [x] I have tested [demo.rerun.io](https://demo.rerun.io/pr/2578) (if applicable)

- [PR Build Summary](https://build.rerun.io/pr/2578)
- [Docs preview](https://rerun.io/preview/pr%3Ajleibs%2Fsave_blueprints/docs)
- [Examples preview](https://rerun.io/preview/pr%3Ajleibs%2Fsave_blueprints/examples)

### CODE REVIEW

File: crates/re_viewer/src/store_hub.rs [hunk 1]
```diff
@@ -141,6 +159,75 @@ impl StoreHub {
...
+        // save the blueprints referenced by `blueprint_by_app_id`, even though
+        // there may be other Blueprints in the Hub.
+        for (app_id, blueprint_id) in &self.blueprint_by_app_id {
+            let blueprint_path = default_blueprint_path(app_id)?;
+            re_log::debug!("Saving blueprint for {app_id} to {:?}", blueprint_path);
+
+            if let Some(blueprint) = self.store_dbs.blueprint(blueprint_id) {
+                // TODO(jleibs): This check isn't perfect. If we GC the blueprint and then insert
+                // more rows we could end up with a collision.
+                if self.blueprint_last_save.get(blueprint_id) != Some(&blueprint.num_rows()) {
```
Comment (jleibs) [inline crates/re_viewer/src/store_hub.rs:176]: @teh-cmc num_rows() is close to what I wanted, but not quite... I struggled a bit coming up with an easy API call to get a unique/stable "most recently inserted row". At the moment I'm leaning towards adding this as a bit of additional book-keeping in the MetadataRegistry, but curious if you have a better suggestion.

Comment (teh-cmc) [reply crates/re_viewer/src/store_hub.rs:176]: Let's chat after standup!

File: crates/re_data_store/src/entity_properties.rs [hunk 2]
```diff
@@ -116,6 +125,30 @@ impl EntityProperties {
                 .clone(),
         }
     }
+
+    /// Determine whether this `EntityProperty` has user-edits relative to another `EntityProperty`
+    /// This is similar in concept to `PartialEq`, but more forgiving of Auto taking on different values.
+    pub fn unedited(&self, other: &Self) -> bool {
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_data_store/src/entity_properties.rs:133]: Turns out this didn't break autovalues... a bug I had in the original implementation of unedited broke autovalues. With that fixed this seems to be working now.

File: crates/re_data_store/src/store_db.rs [hunk 3]
```diff
@@ -249,6 +249,10 @@ impl StoreDb {
             + self.entity_db.data_store.num_temporal_rows() as usize
     }
 
+    pub fn last_insert_id(&self) -> u64 {
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_data_store/src/store_db.rs:252]: None of the other fn's in this impl have docstrings...

File: crates/re_viewer/src/saving.rs [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+#[cfg(not(target_arch = "wasm32"))]
+/// Determine the default path for a blueprint based on its `ApplicationId`
+/// This path should be determnistic and unique.
+// TODO(2579): Implement equivalent for web
+pub fn default_blueprint_path(app_id: &ApplicationId) -> anyhow::Result<std::path::PathBuf> {
+    use std::hash::{BuildHasher, Hash as _, Hasher as _};
+
+    use anyhow::Context;
+
+    // TODO(jleibs) is there a better way to get this folder from egui?
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_viewer/src/saving.rs:16]: https://github.com/emilk/egui/issues/3134

File: crates/re_viewer/src/saving.rs [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+    let ent_op_msgs = store_db
+        .iter_entity_op_msgs()
+        .map(|msg| LogMsg::EntityPathOpMsg(store_db.store_id().clone(), msg.clone()))
+        .collect_vec();
+
+    let time_filter = time_selection.map(|(timeline, range)| {
+        (
+            timeline,
+            TimeRange::new(range.min.floor(), range.max.ceil()),
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_viewer/src/saving.rs:94]: I only moved this function over from `app.rs`; I didn't write it.  I think this is a @teh-cmc question.

File: crates/re_viewer/src/store_hub.rs [hunk 6]
```diff
@@ -19,6 +25,10 @@ pub struct StoreHub {
     application_id: Option<ApplicationId>,
     blueprint_by_app_id: HashMap<ApplicationId, StoreId>,
     store_dbs: StoreBundle,
+
+    // The number of rows in the blueprint the last time it was saved
+    #[cfg(not(target_arch = "wasm32"))]
+    blueprint_last_save: HashMap<StoreId, u64>,
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_viewer/src/store_hub.rs:31]: Missed updating the comment when I was changing from the old (rows) implementation to the new `insertion id`.

File: crates/re_viewer/src/store_hub.rs [hunk 7]
```diff
@@ -141,6 +161,75 @@ impl StoreHub {
...
+        // Because we save blueprints based on their `ApplicationId`, we only
+        // save the blueprints referenced by `blueprint_by_app_id`, even though
+        // there may be other Blueprints in the Hub.
+        for (app_id, blueprint_id) in &self.blueprint_by_app_id {
+            let blueprint_path = default_blueprint_path(app_id)?;
+            re_log::debug!("Saving blueprint for {app_id} to {:?}", blueprint_path);
+
+            if let Some(blueprint) = self.store_dbs.blueprint(blueprint_id) {
+                // TODO(jleibs): This check isn't perfect. If we GC the blueprint and then insert
+                // more rows we could end up with a collision.
```
[the remark this answers is not in the document]
Comment (jleibs) [reply crates/re_viewer/src/store_hub.rs:177]: obsolete comment from when I had implemented on rows.  "Collision" would have been rows being the same due to GC + new insertions, despite the content of the DB being different.

File: crates/re_data_store/src/store_db.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (emilk) [reply crates/re_data_store/src/store_db.rs]: I don't find `last_insert_id` and `u64` self-documenting enough. Of course, YMMV.

If you find other functions cryptic, please add docstrings to them too (or ask their author to do so)

File: Cargo.lock [hunk 8]
```diff
@@ -5020,6 +5022,16 @@ dependencies = [
...
 ]
 
+[[package]]
+name = "sanitize-filename"
+version = "0.4.0"
+source = "registry+https://github.com/rust-lang/crates.io-index"
+checksum = "08c502bdb638f1396509467cb0580ef3b29aa2a45c5d43e5d84928241280296c"
+dependencies = [
+ "lazy_static",
+ "regex",
```
[the remark this answers is not in the document]
Comment (jleibs) [reply Cargo.lock]: What's the best way to see the scoped dependencies for just the web-build.  I had checked that several other packages depended on regex but hadn't realized they were only native. Either way, implemented my own sanitizer.

[the remark this answers is not in the document]
Comment (emilk) [reply Cargo.lock]: I don't know. `cargo tree --target wasm32-unknown-unknown -i -p regex` doesn't work as expected (it includes `env_logger`, despite that being behind `[target.'cfg(not(target_arch = "wasm32"))'.dependencies]`)

File: crates/re_arrow_store/src/store.rs [hunk 9]
```diff
@@ -155,6 +155,10 @@ impl std::ops::DerefMut for ClusterCellCache {
 
 // ---
 
+/// Incremented on each edit
+#[derive(Debug, Default, Clone, PartialEq, Eq, PartialOrd, Ord)]
```
Comment (emilk) [inline crates/re_arrow_store/src/store.rs:159]:
```suggestion
#[derive(Debug, Default, Clone, Hash, PartialEq, Eq, PartialOrd, Ord)]
```

**Review by emilk — APPROVED** [review]

**Pull request merged**